### PR TITLE
add base metadata query (#5954)

### LINKED ISSUE #5954: [Feature] Support metadata management
### 1. Summary

Metadata provides necessary data support for data governance because supporting metadata is very important. It can not only enable users to understand data more clearly when using DS development, such as table fields and field types. It solves the "data search demand" of retrieving and understanding data, and can provide better services for data quality and data security in the later stage. It can also improve the efficiency of data development，I have implemented it and run stably in production. After discussion, I will submit pr

### 2. Key design points**

-Provide unified interface

-Get metadata in real time without external storage

-Insert metadata and run independently

-Add different data sources and put jars into the Lib directory without restarting the dolphin API server and work server

image

**2.1.Installation and deployment directory structure**

├──   Lib lib directory

│     ├──   Meta common -- spring public dependency package directory

│     ├──   Hadoop client -- lib directory on which Hadoop depends

│     ├──   JDBC -- jar package directory necessary for connecting data source with Java

│     ├──   ├---── hive -- jar package necessary to connect hive with Java

│     ├──   ├---──pg

│     ├──   ├---──mysql

│     ├──   Plugin -- get the jar package directory of the corresponding datasourcecilent from the data source and metadata

│     ├──   ├---──hive

│     ├──   ├---──pg

│     ├──   ├---──mysql

**2.2.Source directory structure**

├──   Data source plug-in directory supported by meta plugin

│     ├──   Hive plugin -- hiveclien connection acquisition, metadata function

│     ├── ├──lib - jar package necessary for connecting hive with Java

│     ├──    postgresql-plugin

│     ├── ├──lib

│     ├──    mysql-plugin

│     ├── ├──lib

│     ├──    oracle-plugin

│     ├── ├──lib

│     └──    sqlserver-plugin

│     ├── ├──lib

│     └──    clickhouse-plugin

│     ├── ├──lib

│     └──    db2-plugin

│     ├── ├──lib

├──   meta-api   --provides data source management and its metadata query restful service, and plug-in datasourceclient is created

**2.3.Create data source flowchart**

image

### 3. Demand analysis

For metadata, we must pay attention to scalability and support multiple data sources. Therefore, the integrated design method should be adopted. The following functions need to be met to realize it

##### 3.1.1 create data source

**The current shortcomings and the benefits of improvement are**

Now to add different data sources, you need to put jars into the Lib directory, and then restart dolphin API server and work server. After the expansion is realized, you will no longer need to restart. Later, it is also very helpful for the plug-in of metadata

```

{

"Connhdfs": true, -- hive, etc. need to connect metadata,

"Connmetastore": true, -- hive, etc. need to connect metadata

"Dbtype": "hive", -- data source type

"driverLocation": "/home/bigdata/meta/jdbc/hive/hive-jdbc-2.2.2-standalone.jar",

"Hadoop configurationresources": "/ home / bigdata / apps / hadoop-2.7.2/etc/hadoop/core-site.xml, / home / bigdata / apps / hadoop-2.7.2/etc/hadoop/hdfs-site.xml, / home / bigdata / apps / apache-hive-1.2.1-bin/conf/hive-site.xml", -- get from the configuration file

"id": 1,

"Kerberos KeyTab": "", --- get from configuration file

"Kerberos krb5conf": "", --- get from configuration file

"Principal": "",

"password": 123456,

"URL": "JDBC: hive2: / / zbg-bg-01:10000 / default", -- assembled according to the input parameters, or filled in manually by the user

"username": "root",

"Other": null --- other parameters

}



```

### **4. Data source SPI to do list**

- [x] basic code for SPI implementation of data source and metadata, loading of data source driver and plugin, and unified management of data source connection

- [x] load and transform the existing supported data source SPI, and add the data source client check and SQL execution functions
- [x] Replace the data source connection loaded by SPI where the original connection method is used in DS

###5. metadata acquisition

### PR DESCRIPTION
**Pull request by Narcasserun**
- add base metadata query 
- add hive metadata query

This is close #5954

### CODE REVIEW

**Comment by Narcasserun** [comment]
@caishunfeng owasp check keeps failing, can you help me find the reason?

**Comment by SbloodyS** [comment]
> @caishunfeng owasp check keeps failing, can you help me find the reason?

This is the current situation. You can ignore it first. This does not affect code merging...

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java [hunk 1]
```diff
@@ -101,6 +117,140 @@ public void checkClient() {
...
+        throw new UnsupportedOperationException("NOT_SUPPORT");
+    }
+
+    @Override
+    public List<Map<String, Object>> getTableStruct(String dbName, String schemaName, String tableName) {
+        throw new UnsupportedOperationException("NOT_SUPPORT");
+    }
+
+    @Override
+    public MutableTriple<Map<String, String>, List<Map<String, Object>>, List<Map<String, String>>> executeSql(
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java:143]: Can we return a pojo here?

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java [hunk 2]
```diff
@@ -101,6 +117,140 @@ public void checkClient() {
...
+
+    private String getColumnName(String columnName) {
+        return columnName.contains(".") ? columnName.split("\\.", 2)[1] : columnName;
+    }
+
+    protected MutableTriple<Map<String, String>, List<Map<String, Object>>, List<Map<String, String>>>
+    executeSqlListReturnLast(JdbcTemplate jdbcTemplate, List<String> querySqlList) {
+        Stopwatch stopwatchForAll = Stopwatch.createStarted();
+        try {
+            return jdbcTemplate.execute(
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java:172]: Can we directly use `query` for each given `sql`?

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java [hunk 3]
```diff
@@ -46,34 +47,38 @@ public class DataSourceClientProvider {
...
-            }
-            return dataSourceChannel.createDataSourceClient(baseConnectionParam, dbType);
-        });
-        return dataSourceClient.getConnection();
+        try {
+            return uniqueId2dataSourceClientCache.get(datasourceUniqueId, () -> {
+                Map<String, DataSourceChannel> dataSourceChannelMap = dataSourcePluginManager.getDataSourceChannelMap();
+                DataSourceChannel dataSourceChannel = dataSourceChannelMap.get(dbType.getDescp());
+                if (null == dataSourceChannel) {
+                    throw DataSourceException.getInstance(String.format("datasource plugin '%s' is not found", dbType.getDescp()));
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java]:
```suggestion
                    throw new DataSourceException(String.format("datasource plugin '%s' is not found", dbType.getDescp()));
```
Directly new an exception is better?

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-hive/src/main/java/org/apache/dolphinscheduler/plugin/datasource/hive/MetaStoreClientPool.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+ * a new client is created and added to the pool. The idle pool can expand till a maximum
+ * size of MAX_HMS_CONNECTION_POOL_SIZE, beyond which the connections are closed.
+ *
+ * This default implementation reads the Hive metastore configuration from the HiveConf
+ * object passed in the c'tor. If you are looking for a temporary HMS instance created
+ * from scratch for unit tests, refer to EmbeddedMetastoreClientPool class. It mocks an
+ * actual HMS by creating a temporary Derby backend database on the fly. It should not
+ * be used for production Catalog server instances.
+ */
+public class MetaStoreClientPool {
```
Comment (ruanwenjun) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-hive/src/main/java/org/apache/dolphinscheduler/plugin/datasource/hive/MetaStoreClientPool.java:29]: Do we need to add license about this file, it seems this file is very similar in other project.

Comment (Narcasserun) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-hive/src/main/java/org/apache/dolphinscheduler/plugin/datasource/hive/MetaStoreClientPool.java:29]: how to add license?

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java (hunk 3 above)
Comment (Narcasserun) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java]: They seem to achieve the same

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java (hunk 1 above)
Comment (Narcasserun) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java:143]: There will be other definitions for pojo

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-hive/src/main/java/org/apache/dolphinscheduler/plugin/datasource/hive/MetaStoreClientPool.java (hunk 4 above)
Comment (ruanwenjun) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-hive/src/main/java/org/apache/dolphinscheduler/plugin/datasource/hive/MetaStoreClientPool.java:29]: You can reference `CodeGenerateUtils`, you need to add announce in PROJECT_HOME/LICENSE file.

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java (hunk 3 above)
Comment (ruanwenjun) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java]: But can reduce a method call?

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java [hunk 5]
```diff
@@ -101,6 +117,140 @@ public void checkClient() {
...
+        return Lists.asList(switchEnvironment(dbName, schemaName), querySql.split(";")).stream()
+            .filter(StringUtils::isNotBlank)
+            .collect(Collectors.toList());
+    }
+
+    private String getColumnName(String columnName) {
+        return columnName.contains(".") ? columnName.split("\\.", 2)[1] : columnName;
+    }
+
+    protected MutableTriple<Map<String, String>, List<Map<String, Object>>, List<Map<String, String>>>
```
Comment (caishunfeng) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java:168]: It is necessary to add some notes and example to explain the  `MutableTriple`.

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java [hunk 6]
```diff
@@ -101,6 +117,140 @@ public void checkClient() {
...
+    executeSqlListReturnLast(JdbcTemplate jdbcTemplate, List<String> querySqlList) {
```
Comment (caishunfeng) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/client/CommonDataSourceClient.java:169]: What's the meaning of `return last`?

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java (hunk 3 above)
Comment (caishunfeng) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/plugin/DataSourceClientManager.java]: >Directly new an exception is better?

+1

**Review by caishunfeng — COMMENTED** [review]
just a quick review.

**Pull request closed without merging**